### PR TITLE
Implementation of Secured Multi-Party Linear Regression

### PR DESCRIPTION
**Pull request by andrelmfarias**
Implementation of Encrypted Linear Regression using MPC. This implementation follows the algorithm in section 2 of [Jonathan Bloom's paper](https://arxiv.org/pdf/1901.09531.pdf)

I did some tests locally and they are passing, although the training is considerably slow (mainly due to secure division). For example, for 3 covariates the train runs for 1min and for 10 covariates it takes around 6min30s.

For now, the implementation uses a random worker in the pool of workers, a crypto provider worker and an "Honest but Curious" worker in order to perform 2-party computation (current constraint created by secure comparison inside secure division).

TO DO's:

- [x] Implement unit tests
- [x] Cases for `fit_intercept`
- [x] Compute p-values 
- [x] Implement `.summarize()` method
- [x] Implement `.predict()` method
- [x] Improve documentation

### CODE REVIEW

File: syft/frameworks/torch/linalg/lr.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+    def fit(self, X_ptrs: List[torch.Tensor], y_ptrs: List[torch.Tensor]):
+        """
+        Fits the linear model using Secured Multi-Party Linear Regression.
+        The final results (i.e. coefficients and p-values) will be public.
+        """
+
+        # Checking if the pointers are as expected
+        self._check_ptrs(X_ptrs, y_ptrs)
+
+        # Check if each y is a 2-dim or 1-dim tensor, unsqueeze if it is not
```
Comment (mari-linhares) [inline syft/frameworks/torch/linalg/lr.py]: Is this comment correct?

Comment (andrelmfarias) [reply syft/frameworks/torch/linalg/lr.py]: Yes, it's a bit ambiguous though...

I meant "Check if each y is a 2-dim or 1-dim tensor, unsqueeze if it 1-dim"

**Comment by andrelmfarias** [comment]
> If you are willing to add more links to the paper in the fit and _compute_pvalues methods (as you have made in other parts of the code), I promise I will review it as well

For the fit method, you can take a look at Section 2 of [Bloom's paper](https://arxiv.org/pdf/1901.09531.pdf) (it's much easier to understand that the ones about crypto protocols we were discussing 😃) and at this [demo notebook](https://github.com/jbloom22/DASH/blob/master/multiparty_linear_regression.ipynb) I pushed to his repo. You can also find the p-values formula in the notebook, which I took from [Bloom's script in R](https://github.com/jbloom22/DASH/blob/master/dash.r). 

The idea p-value computation is that under the hypothesis H0 that a coefficient is 0, the t-statistic `coeff / sigma` follows a Student's t-distribution. The pvalue is the probability of getting a larger value than our empirical computed value under the hypothesis that the "true value" is 0, we compute it by using the value of the cumulated probability at the tails of the distribution (`2 * t.cdf(statistic)`. I did my best to explain it here, but it's much better and intuitive to do it on a board 😞 ...

> except that you should also add the case fit_intercept=False in your test. (you can use @pytest.mark.parametrize)

Will do

File: syft/frameworks/torch/linalg/lr.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        It also computes parallelly  some Regressor's attributes such as number of features and
+        total sample size.
+        """
+        # Set number of features
+        self.n_features = X_ptrs[0].shape[1]
+
+        x_size, y_size = 0, 0
+        for x, y in zip(X_ptrs, y_ptrs):
+            # Check wrapper
+            if not (x.has_child and y.has_child):
```
[the remark this answers is not in the document]
Comment (andrelmfarias) [reply syft/frameworks/torch/linalg/lr.py]: thanks for that one! Done!

File: syft/frameworks/torch/linalg/lr.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        self._check_ptrs(X_ptrs, y_ptrs)
+
+        # Check if each y is a 2-dim or 1-dim tensor, unsqueeze it if it's 1-dim
+        for i, y in enumerate(y_ptrs):
+            if len(y.shape) < 2:
+                y_ptrs[i] = y.unsqueeze(1)
+
+        if self.fit_intercept:
+            self._add_intercept(X_ptrs)
+            self._dgf -= 1
```
[the remark this answers is not in the document]
Comment (andrelmfarias) [reply syft/frameworks/torch/linalg/lr.py]: Yes it's exactly degrees of freedom (N - K in the paper)

[the remark this answers is not in the document]
Comment (andrelmfarias) [reply syft/frameworks/torch/linalg/lr.py]: Actually there's already a line comment about it (when the attribute `self._dgf` is initialized):

https://github.com/OpenMined/PySyft/blob/b4229455fc226001b61cb06bd204d8c47776cdf8/syft/frameworks/torch/linalg/lr.py#L209-L210

**Pull request merged**